Expose getaddrinfo errors. Was: Return empty list on EAI_ errors for getaddrinfo@luv (#351)

haesbaert opened the pull request:
Unix.getaddrinfo used in u-ring ignores gar_errno and returns an empty list, this makes the luv backend follow a similar behaviour.

Should fix #351 where we would get EAI_NONAME (or others) if resolution failed, this is a bit defensive as we just handle expected EAI errors.

haesbaert committed:
Return empty list on EAI_ errors for getaddrinfo@luv (#351)

Unix.getaddrinfo used in u-ring ignores gar_errno and returns an empty list, this
makes the luv backend follow a similar behaviour.

haesbaert commented:
cc @patricoferris

talex5 commented on lib_eio_luv/eio_luv.ml:
```diff
@@ -575,10 +575,15 @@ module Low_level = struct
+             (fun loop -> Luv.DNS.getaddrinfo ~loop ~request ~service ~node ()))
+    with
+    | Ok nl -> List.filter_map to_eio_sockaddr_t nl
+    | Error `EAI_ADDRFAMILY | Error `EAI_AGAIN | Error `EAI_BADFLAGS | Error `EAI_BADHINTS
+    | Error `EAI_CANCELED   | Error `EAI_FAIL  | Error `EAI_FAMILY   | Error `EAI_MEMORY
+    | Error `EAI_NODATA     | Error `EAI_NONAME| Error `EAI_OVERFLOW | Error `EAI_PROTOCOL
+    | Error `EAI_SERVICE    | Error `EAI_SOCKTYPE -> []
```
This list seems excessive (even if it does match the stdlib version). In particular, `EAI_AGAIN`, `EAI_BADFLAGS`, `EAI_MEMORY` and `EAI_SOCKTYPE` seem like they should be reported as errors, rather than claiming there are no addresses. Maybe `EAI_FAIL` too.

haesbaert replied:
I thought about it too, just some notes:
EAI_FAIL and EAI_AGAIN can be the result of a an actual DNS reply (would have to dig deeper, but those should return an empty list). EAI_SOCKTYPE, EAI_BADFLAGS, EAI_FAMILY we could, since they are mostly programming errors from our side.
I don't have a strong opinion, but I'd err on the side of just emulating the same behaviour of `Unix.getaddrinfo` which ignores everything.

talex5 replied:
I wonder if we should just copy the `getaddrinfo` API and not allow an empty result. The wrapper function could enforce this:

```ocaml
let getaddrinfo ?(service="") (t:#t) hostname =
  match t#getaddrinfo ~service hostname with
  | [] -> raise (Dns_error (Failure "No addresses for ..."))
  | xs -> xs
```

I guess most people getting an empty list will want to report an error, and it's a bit silly you have to guess what the error was when we just threw the real error away. e.g. in `with_tcp_connect` we currently do:

```ocaml
    | [] -> raise (Connection_failure (Failure (Fmt.str "No TCP addresses for %S" host)))
```
Then we could say if that's because the DNS server is busy, the entry doesn't exist, doesn't have an IP address, etc.

haesbaert replied:
Also `Dns_error` is a bit imprecise since `getaddrinfo` is not only about DNS (`AI_NUMERICHOST`and so on)

haesbaert replied:
Thinking a bit more about this, I think `connect` should take a `Sockaddr.stream list`, and loop over the candidates until it succeeds connecting to one. This is virtually what everyone does with `getaddrinfo` + `connect` and `bind`.
In this case `connect` itself could raise something, `getaddrinfo` could still return just an empty list, and the user would still have to handle an exception.

talex5 replied:
Yes, the error code was just an example. We can use a better exception name and contents.

> I think connect should take a Sockaddr.stream list, and loop over the candidates until it succeeds connecting to one.

`with_tcp_connect` already handles looping over the candidates; I'm not sure there's much benefit to changing `connect` too. In fact, that would just mean having to do something for the extra case of an empty list.

> In this case connect itself could raise something, getaddrinfo could still return just an empty list, and the user would still have to handle an exception.

I don't see how that helps. We'd still be losing the detailed error from `getaddrinfo` and replacing it with a guess.

haesbaert replied:
can I get a ruling here ?

talex5 replied:
> Should we just raise all errors from `getaddrinfo` on both cases?

Yes, that makes sense to me.

( Having `connect` take a `Sockaddr.stream list` doesn't seem useful to me. You can't give a good error if there are no addresses - you don't know what the DNS error was, what the failing hostname was, or even whether DNS was used at all! )

talex5 commented:
> I'll differ to @talex5's knowledge of which EAI_ errors to catch and which to no :))

I have no special knowledge - just reading the `getaddrinfo` man-page :-)

haesbaert committed:
works

haesbaert committed:
EINVAL on bad EAI

haesbaert committed:
Move things to net and implement libuv

haesbaert committed:
Move getaddrinfo stubs into its own file due to copyrights

haesbaert committed:
zap gai_strerror bindings, copy strings from libc

haesbaert committed:
Add Net.getaddrinfo_to_string

haesbaert committed:
Add EAI_PROTOCOL and don't be sloppy with defines

haesbaert committed:
restore eio_stubs properly now that we moved getaddrinfo

haesbaert committed:
whitespace

haesbaert committed:
EAI_PROTOCOL WHOOPS

haesbaert committed:
Add EAI_NONAME tests

haesbaert committed:
EAI_OVERFLOW

haesbaert committed:
Document semantics in net.mli

haesbaert committed:
Better paranoia

haesbaert committed:
We can't really guarantee the error code since it depends on the environment.

haesbaert commented:
Some corrections to my previous statement, getaddrinfo(3) *cannot* return an empty list and no error, it's either something or an error. I checked libc.

This PR should be ready for reviewing, some notes:
- A lot of the stub is just copied from ocaml's stdlib
- The values chosen for the undefined `EAI_*` values are taken from libuv (since it works for them it works for us), they're added for completion sakeness and a future guard. I'm sure once we have some other backend those will popup and we will move the stub, or maybe glibc adds them in the future, who knows.
- The strings for `getaddrinfo_error_to_string` were copied verbatim from libc.

haesbaert committed:
fix tests

haesbaert commented:
Tests are zefixed, marking this as ready for review

haesbaert committed:
zap duplicate

talex5 commented on lib_eio_linux/getaddrinfo_stubs.c:
```diff
@@ -0,0 +1,187 @@
+/*     en Automatique.                                                    */
+/*   Copyright 2022 Tarides                                               */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
```
I think we want to avoid LGPL code in Eio (it's currently BSD and ISC only). Though they might relicense this bit if asked.

haesbaert replied:
@gasche @xavierleroy (sorry for the ping).
I've copied significant part of `otherlibs/unix/getaddrinfo.c` into EIO (https://github.com/ocaml-multicore/eio/blob/61b737d5be683b9be611844b16c7e73b5bece09a/lib_eio_linux/getaddrinfo_stubs.c). The original file is LGPLed and we would like to relicense our new file to ISC.
Do you permit us to relicense the bits we copied out from LGPL to ISC in EIO ?

xavierleroy replied:
No.  I only use copyleft licenses for my free software.  Please consider using the LGPL with OCaml linking exception for your library, or do not reuse any of my code.

haesbaert replied:
> No. I only use copyleft licenses for my free software. Please consider using the LGPL with OCaml linking exception for your library, or do not reuse any of my code.

Thank you, will do one of the two as suggested.

talex5 reviewed (commented):
Looks sensible.

Note that I'm changing the error handling quite a lot in #378, which will impact this PR a bit.

dra27 commented on lib_eio_linux/getaddrinfo_stubs.c:
```diff
@@ -0,0 +1,187 @@
+#include <caml/unixsupport.h>
+#include <caml/socketaddr.h>
+
+static value caml_unix_cst_to_constr(int n, int *tbl, int size, int deflt)
```
You could avoid this duplication by borrowing the `extern` declaration in `cst2constr.h`

haesbaert replied:
fixed, somehow my brain remembered the function being static in first place.

dra27 commented on lib_eio_linux/getaddrinfo_stubs.c:
```diff
@@ -0,0 +1,187 @@
+  if (retcode == 0) {
+    for (r = res; r != NULL; r = r->ai_next) {
+      v = caml_alloc_small(2, Tag_cons);
+      Field(v, 0) = convert_addrinfo(r);
```
This violates GC rule 5 - the `convert_addrinfo` call needs to be done before the `caml_alloc_small` and stored in a registered local (`convert_addrinfo` allocates).

haesbaert replied:
thanks for spotting this, fixed it now

dra27 reviewed (commented):
Drive-by comments on the borrowed C code!

haesbaert committed:
Just extern the declaration so we don't need to copy cst_to_constr.

Many thanks to dra27 !

Somehow this function was static on first place in my head.

haesbaert committed:
Address dra27 point #2, many thanks again !

When I removed the `e` and I violated GC #rule 5. Basically a small block must
be fully initialized before we trigger the next allocation, TIL.

talex5 commented:
I updated this code (with fresh C stubs) and it's now merged as #866 - thanks!
